File: sdks/python/src/fireweave/openfeature/local_provider.py
"""``make_fireweave_local_provider()`` — OpenFeature provider for the DEV branch.

Wires :class:`FireweaveLocalAdapter` through the ordinary
:class:`FireweaveRuntime` + :class:`FireweaveProvider` stack, then applies one
narrow rewrite on the way out: ``FLAG_NOT_FOUND`` becomes a clean ``DEFAULT``
resolution. Real defects (``PROVIDER_NOT_READY``, ``INVALID_CONTEXT``,
``TYPE_MISMATCH``, ``PROVIDER_FATAL``) pass through untouched.
"""

from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any, Callable, List, Mapping, Optional, Sequence, Union

from openfeature.evaluation_context import EvaluationContext as OFContext
from openfeature.exception import ErrorCode
from openfeature.flag_evaluation import FlagResolutionDetails, Reason as OFReason
from openfeature.provider import AbstractProvider, Metadata

from ..adapters.local import FireweaveLocalAdapter
from ..runtime import FireweaveRuntime
from .provider import FireweaveProvider

__all__ = [
    "FwLocalCapture",
    "get_fw_local_captures",
    "reset_fw_local_captures",
    "make_fireweave_local_provider",
]


@dataclass(frozen=True)
class FwLocalCapture:
    flag_key: str
    type: str  # boolean | string | integer | float | object | number
    value: Any
    reason: str
    ts: float


_captures: List[FwLocalCapture] = []


def get_fw_local_captures() -> Sequence[FwLocalCapture]:
    """Every evaluation observed through a local provider in this process."""
    return tuple(_captures)


def reset_fw_local_captures() -> None:
    """Clear the capture buffer (call between tests)."""
    global _captures
    _captures = []


class _FireweaveLocalProvider(AbstractProvider):
    def __init__(
        self,
        *,
        dev_flags: Optional[Mapping[str, bool]] = None,
        echo: bool = False,
        now: Optional[Callable[[], float]] = None,
    ) -> None:
        adapter = FireweaveLocalAdapter(dev_flags=dev_flags)
        # Eager readiness: a laptop has nothing to connect to.
        self._inner = FireweaveProvider(FireweaveRuntime(adapter))
        self._echo = echo
        self._now = now or time.time

    def get_metadata(self) -> Metadata:
        return Metadata(name="fireweave-local")

    def initialize(self, evaluation_context: OFContext) -> None:
        self._inner.initialize(evaluation_context)

    def shutdown(self) -> None:
        self._inner.shutdown()

    def resolve_boolean_details(
        self,
        flag_key: str,
        default_value: bool,
        evaluation_context: Optional[OFContext] = None,
    ) -> FlagResolutionDetails[bool]:
        return self._finish(
            "boolean",
            flag_key,
            default_value,
            self._inner.resolve_boolean_details(
                flag_key, default_value, evaluation_context
            ),
        )

    def resolve_string_details(
        self,
        flag_key: str,
        default_value: str,
        evaluation_context: Optional[OFContext] = None,
    ) -> FlagResolutionDetails[str]:
        return self._finish(
            "string",
            flag_key,
            default_value,
            self._inner.resolve_string_details(
                flag_key, default_value, evaluation_context
            ),
        )

    def resolve_integer_details(
        self,
        flag_key: str,
        default_value: int,
        evaluation_context: Optional[OFContext] = None,
    ) -> FlagResolutionDetails[int]:
        return self._finish(
            "integer",
            flag_key,
            default_value,
            self._inner.resolve_integer_details(
                flag_key, default_value, evaluation_context
            ),
        )

    def resolve_float_details(
        self,
        flag_key: str,
        default_value: float,
        evaluation_context: Optional[OFContext] = None,
    ) -> FlagResolutionDetails[float]:
        return self._finish(
            "float",
            flag_key,
            default_value,
            self._inner.resolve_float_details(
                flag_key, default_value, evaluation_context
            ),
        )

    def resolve_object_details(
        self,
        flag_key: str,
        default_value: Union[Sequence[Any], Mapping[str, Any]],
        evaluation_context: Optional[OFContext] = None,
    ) -> FlagResolutionDetails[Union[Sequence[Any], Mapping[str, Any]]]:
        return self._finish(
            "object",
            flag_key,
            default_value,
            self._inner.resolve_object_details(
                flag_key, default_value, evaluation_context
            ),
        )

    def _finish(
        self,
        type_: str,
        flag_key: str,
        default_value: Any,
        details: FlagResolutionDetails[Any],
    ) -> FlagResolutionDetails[Any]:
        if details.error_code == ErrorCode.FLAG_NOT_FOUND:
            resolved = FlagResolutionDetails(
                value=default_value,
                variant="default",
                reason=OFReason.DEFAULT,
            )
        else:
            resolved = details
        self._record(
            type_,
            flag_key,
            resolved.value,
            str(resolved.reason) if resolved.reason is not None else "UNKNOWN",
        )
        return resolved

    def _record(self, type_: str, flag_key: str, value: Any, reason: str) -> None:
        _captures.append(
            FwLocalCapture(
                flag_key=flag_key,
                type=type_,
                value=value,
                reason=reason,
                ts=self._now(),
            )
        )
        if self._echo:
            print(
                f"[fw-local] {type_} {flag_key} = {json.dumps(value, default=str)} ({reason})"
            )


def make_fireweave_local_provider(
    *,
    dev_flags: Optional[Mapping[str, bool]] = None,
    echo: bool = False,
    now: Optional[Callable[[], float]] = None,
) -> AbstractProvider:
    """Build the dev-branch OpenFeature provider.

    Example::

        from fireweave.openfeature import make_fireweave_local_provider

        provider = make_fireweave_local_provider(
            echo=True,
            dev_flags={"my-feature": True},
        )
    """
    return _FireweaveLocalProvider(dev_flags=dev_flags, echo=echo, now=now)

File: sdks/node/packages/sdk/src/local-provider.ts
/**
 * `makeFireweaveLocalProvider()` — the OpenFeature provider a scaffolded
 * harness binds on its DEV branch.
 *
 * It wires {@link FireweaveLocalAdapter} through the ordinary
 * {@link FireweaveRuntime} + {@link FireweaveProvider} stack, then applies one
 * narrow rewrite on the way out.
 *
 * ## Why the rewrite exists
 *
 * The runtime turns an adapter miss into an ERROR decision carrying
 * `FLAG_NOT_FOUND`. That is correct for a real backend — asking for a control
 * point the server has never heard of is a genuine misconfiguration — and it
 * is pinned across all four language SDKs by
 * `contracts/evaluation/eval-missing-flag-default.json`, so it is a ratified
 * contract rather than an implementation detail.
 *
 * On the dev substrate the same signal means the opposite. "Not configured
 * locally" is the *normal* state of nearly every control point on a laptop:
 * the harness scaffolds both branches up front, and call sites default to
 * `false` until a rollout ramps them. Surfacing that as an error would make
 * every dev read an error resolution — which pollutes OpenFeature error hooks,
 * inflates flag-error telemetry for anyone running the OTel flag hooks, and
 * buries the errors that do matter.
 *
 * So exactly one outcome is rewritten: `FLAG_NOT_FOUND` becomes a clean
 * `DEFAULT` resolution carrying the caller's own default value. Everything
 * else — `PROVIDER_NOT_READY`, `INVALID_CONTEXT`, `TYPE_MISMATCH`,
 * `PROVIDER_FATAL` — passes through untouched, because each of those is a real
 * defect in dev too, and a dev provider that hides them is worse than no dev
 * provider.
 *
 * ## Observability
 *
 * Two ways to see what happened, mirroring the pair the harness already
 * expects: humans read the optional `echo` line; tests read
 * {@link getFwLocalCaptures}. Captures are recorded here rather than in the
 * adapter so each entry reflects what the caller actually observed — including
 * the rewritten `DEFAULT` reason, which the adapter never sees.
 */
import {
  ErrorCode,
  StandardResolutionReasons,
  type EvaluationContext,
  type JsonValue,
  type Provider,
  type ProviderMetadata,
  type ResolutionDetails,
} from '@openfeature/server-sdk';
import { FireweaveProvider } from './provider.js';
import { FireweaveRuntime } from './runtime.js';
import { FireweaveLocalAdapter } from './adapters/local.js';

export interface FwLocalCapture {
  readonly flagKey: string;
  readonly type: 'boolean' | 'string' | 'number' | 'object';
  readonly value: unknown;
  readonly reason: string;
  readonly ts: number;
}

let captures: FwLocalCapture[] = [];

/** Every evaluation observed through a local provider in this process. */
export function getFwLocalCaptures(): readonly FwLocalCapture[] {
  return captures;
}

/** Clear the capture buffer (call between tests). */
export function resetFwLocalCaptures(): void {
  captures = [];
}

export interface FireweaveLocalProviderOptions {
  /**
   * Per-key boolean overrides — the only supported way to turn a control point
   * ON (or force it OFF) locally. Absent keys resolve to the call-site default.
   */
  readonly devFlags?: Record<string, boolean>;
  /** Print one line per evaluation to the console. Default `false`. */
  readonly echo?: boolean;
  /** Injected clock, so capture timestamps are deterministic under test. */
  readonly now?: () => number;
}

class FireweaveLocalProvider implements Provider {
  readonly metadata: ProviderMetadata = Object.freeze({ name: 'fireweave-local' });
  readonly runsOn = 'server' as const;

  private readonly inner: FireweaveProvider;
  private readonly echo: boolean;
  private readonly now: () => number;

  constructor(options: FireweaveLocalProviderOptions = {}) {
    const adapterOptions =
      options.devFlags !== undefined ? { devFlags: options.devFlags } : {};
    this.inner = new FireweaveProvider(
      new FireweaveRuntime(new FireweaveLocalAdapter(adapterOptions)),
      // Eager readiness: a laptop has nothing to connect to, so there is no
      // reason to serve NotReady decisions during a startup window that does
      // not exist.
      { lazyReady: false },
    );
    this.echo = options.echo ?? false;
    this.now = options.now ?? (() => Date.now());
  }

  initialize(context?: EvaluationContext): Promise<void> {
    return this.inner.initialize(context);
  }

  onClose(): Promise<void> {
    return this.inner.onClose();
  }

  onContextChange(oldContext: EvaluationContext, newContext: EvaluationContext): Promise<void> {
    return this.inner.onContextChange(oldContext, newContext);
  }

  async resolveBooleanEvaluation(
    flagKey: string,
    defaultValue: boolean,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<boolean>> {
    return this.finish(
      'boolean',
      flagKey,
      defaultValue,
      await this.inner.resolveBooleanEvaluation(flagKey, defaultValue, context),
    );
  }

  async resolveStringEvaluation(
    flagKey: string,
    defaultValue: string,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<string>> {
    return this.finish(
      'string',
      flagKey,
      defaultValue,
      await this.inner.resolveStringEvaluation(flagKey, defaultValue, context),
    );
  }

  async resolveNumberEvaluation(
    flagKey: string,
    defaultValue: number,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<number>> {
    return this.finish(
      'number',
      flagKey,
      defaultValue,
      await this.inner.resolveNumberEvaluation(flagKey, defaultValue, context),
    );
  }

  async resolveObjectEvaluation<T extends JsonValue>(
    flagKey: string,
    defaultValue: T,
    context: EvaluationContext,
  ): Promise<ResolutionDetails<T>> {
    return this.finish(
      'object',
      flagKey,
      defaultValue,
      await this.inner.resolveObjectEvaluation<T>(flagKey, defaultValue, context),
    );
  }

  /** Apply the not-found rewrite, then record what the caller will see. */
  private finish<T>(
    type: FwLocalCapture['type'],
    flagKey: string,
    defaultValue: T,
    details: ResolutionDetails<T>,
  ): ResolutionDetails<T> {
    const resolved =
      details.errorCode === ErrorCode.FLAG_NOT_FOUND
        ? {
            value: defaultValue,
            variant: 'default',
            reason: StandardResolutionReasons.DEFAULT,
          }
        : details;

    this.record(type, flagKey, resolved.value, resolved.reason ?? StandardResolutionReasons.UNKNOWN);
    return resolved;
  }

  private record(
    type: FwLocalCapture['type'],
    flagKey: string,
    value: unknown,
    reason: string,
  ): void {
    captures.push({ flagKey, type, value, reason, ts: this.now() });
    if (this.echo) {
      // eslint-disable-next-line no-console
      console.log(`[fw-local] ${type} ${flagKey} = ${JSON.stringify(value)} (${reason})`);
    }
  }
}

/**
 * Build the dev-branch OpenFeature provider.
 *
 * ```ts
 * // fw-providers.ts (scaffolded by /fireweave:initialise)
 * export function makeDevProvider(): Provider {
 *   return makeFireweaveLocalProvider({
 *     echo: true,
 *     devFlags: { 'my-feature': true },   // dogfood one control point ON
 *   });
 * }
 * ```
 */
export function makeFireweaveLocalProvider(
  options: FireweaveLocalProviderOptions = {},
): Provider {
  return new FireweaveLocalProvider(options);
}
